### PR TITLE
Migrate JSONArrayTest to Kotlin

### PR DESCRIPTION
**Pull request by oyeraghib**
## Purpose / Description
Migrate JSONArrayTest to Kotlin

## Checklist

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Review by david-allison — CHANGES_REQUESTED** [review]
Issue 1: missing the line at the start of the file: `// noinspection MissingCopyrightHeader`

Issue 2: See 

https://github.com/ankidroid/Anki-Android/blob/2f1b2419a0ff8e652c9fcee0efc09891b404847c/lint-rules/src/main/java/com/ichi2/anki/lint/rules/JUnitNullAssertionDetector.kt#L57-L59

<details>

```
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt:69: Error: Use kotlin.test.assert[Not]Null OR kotlin.test.junit.JUnitAsserter instead of JUnit in Kotlin [LegacyNullAssertionDetector from com.ichi2.anki:lint-rules]
        assertNull(array.toJSONObject(JSONArray()))
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt:439: Error: Use kotlin.test.assert[Not]Null OR kotlin.test.junit.JUnitAsserter instead of JUnit in Kotlin [LegacyNullAssertionDetector from com.ichi2.anki:lint-rules]
        assertNull(array.toString())
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt:1: Error: All files in AnkiDroid must contain a GPLv3-compatible copyright header [MissingCopyrightHeader from com.ichi2.anki:lint-rules]
```

</details>

File: AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt [hunk 1]
```diff
@@ -20,507 +20,539 @@
...
-        assertTrue(a.equals(b));
-        assertEquals(a.hashCode(), b.hashCode());
-        b.put(true);
-        assertFalse(a.equals(b));
-        assertTrue(a.hashCode() != b.hashCode());
+    fun testEqualsAndHashCode() {
+        val a = JSONArray()
+        val b = JSONArray()
+        assertTrue(a == b)
+        assertEquals("equals() not consistent with hashCode()", a.hashCode().toLong(), b.hashCode().toLong())
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt:75]: remove both `.toLong()` calls

File: AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt [hunk 2]
```diff
@@ -20,507 +20,539 @@
...
+        a.put(true)
+        a.put(false)
+        b.put(true)
+        b.put(false)
+        assertTrue(a == b)
+        assertEquals(a.hashCode().toLong(), b.hashCode().toLong())
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt:81]: ditto

File: AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt [hunk 3]
```diff
@@ -20,507 +20,539 @@
...
-        assertEquals(false, other.getBoolean(3));
+    fun testBooleans() {
+        val array = JSONArray()
+        array.put(true)
+        array.put(false)
+        array.put(2, false)
+        array.put(3, false)
+        array.put(2, true)
+        assertEquals("[true,false,true,false]", array.toString())
+        assertEquals(4, array.length().toLong())
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt:96]: ditto

File: AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt [hunk 4]
```diff
@@ -20,507 +20,539 @@
...
-        JSONArray array = new JSONArray();
-        array.put(3, null);
-        array.put(0, JSONObject.NULL);
-        assertEquals(4, array.length());
-        assertEquals("[null,null,null,null]", array.toString());
+    fun testNulls() {
+        val array = JSONArray()
+        array.put(3, null)
+        array.put(0, JSONObject.NULL)
+        assertEquals(4, array.length().toLong())
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt:140]: ditto

File: AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt [hunk 5]
```diff
@@ -20,507 +20,539 @@
...
-        array.put(Double.MAX_VALUE);
-        array.put(-0d);
-        assertEquals(4, array.length());
+    fun testNumbers() {
+        val array = JSONArray()
+        array.put(Double.MIN_VALUE)
+        array.put(9223372036854775806L)
+        array.put(Double.MAX_VALUE)
+        array.put(-0.0)
+        assertEquals(4, array.length().toLong())
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt:204]: ditto

File: AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt [hunk 6]
```diff
@@ -20,507 +20,539 @@
...
+        assertEquals(Double.MAX_VALUE, array.getDouble(2), 0.0)
+        assertEquals(-0.0, array.getDouble(3), 0.0)
+        assertEquals(0, array.getLong(0))
+        assertEquals(9223372036854775806L, array.getLong(1))
+        assertEquals(Long.MAX_VALUE, array.getLong(2))
+        assertEquals(0, array.getLong(3))
+        assertEquals(0, array.getInt(0).toLong())
+        assertEquals(-2, array.getInt(1).toLong())
+        assertEquals(Int.MAX_VALUE.toLong(), array.getInt(2).toLong())
+        assertEquals(0, array.getInt(3).toLong())
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt:222]: ditto

File: AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt [hunk 7]
```diff
@@ -20,507 +20,539 @@
...
+        assertEquals(Double.MIN_VALUE, array.opt(0))
+        assertEquals(Double.MIN_VALUE, array.optDouble(0), 0.0)
+        assertEquals(0, array.optLong(0, 1L))
+        assertEquals(0, array.optInt(0, 1).toLong())
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt:226]: ditto

File: AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt [hunk 8]
```diff
@@ -20,507 +20,539 @@
...
+        assertEquals(5, array.optInt(1, 3).toLong())
         // The last digit of the string is a 6 but getLong returns a 7. It's probably parsing as a
         // double and then converting that to a long. This is consistent with JavaScript.
-        assertEquals(9223372036854775807L, array.getLong(2));
-        assertEquals(9.223372036854776E18, array.getDouble(2), 0);
-        assertEquals(Integer.MAX_VALUE, array.getInt(2));
-        assertFalse(array.isNull(3));
+        assertEquals(9223372036854775807L, array.getLong(2))
+        assertEquals(9.223372036854776E18, array.getDouble(2), 0.0)
+        assertEquals(Int.MAX_VALUE.toLong(), array.getInt(2).toLong())
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt:282]: ditto

File: AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt [hunk 9]
```diff
@@ -20,507 +20,539 @@
...
-        values.put(0, 11.0d);
-        assertEquals(5.5d, object.get("a"));
+    fun testToJSONObject() {
+        val keys = JSONArray()
+        keys.put("a")
+        keys.put("b")
+        val values = JSONArray()
+        values.put(5.5)
+        values.put(false)
+        val `object` = values.toJSONObject(keys)
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt:330]: cleanup: have `toJSONObject` return non-null

File: AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt [hunk 10]
```diff
@@ -20,507 +20,539 @@
...
      */
     @Test
-    public void testCreateWithUnsupportedNumbers() {
-        JSONArray array = new JSONArray(Arrays.asList(5.5, Double.NaN));
-        assertEquals(2, array.length());
-        assertEquals(5.5, array.getDouble(0), 0);
-        assertEquals(Double.NaN, array.getDouble(1), 0);
+    fun testCreateWithUnsupportedNumbers() {
+        val array = JSONArray(Arrays.asList(5.5, Double.NaN))
+        assertEquals(2, array.length().toLong())
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt:428]: toLong

File: AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt [hunk 11]
```diff
@@ -20,507 +20,539 @@
...
     }
+
     @Test
-    public void testTokenerConstructor() {
-        JSONArray object = new JSONArray(new JSONTokener("[false]"));
-        assertEquals(1, object.length());
-        assertEquals(false, object.get(0));
+    fun testTokenerConstructor() {
+        val `object` = JSONArray(JSONTokener("[false]"))
+        assertEquals(1, `object`.length().toLong())
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt:451]: toLong

File: AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt [hunk 12]
```diff
@@ -20,507 +20,539 @@
...
     @Test
-    public void testCreate() {
-        JSONArray array = new JSONArray(Arrays.asList(5.5, true));
-        assertEquals(2, array.length());
-        assertEquals(5.5, array.getDouble(0), 0);
-        assertEquals(true, array.get(1));
-        assertEquals("[5.5,true]", array.toString());
+    fun testCreate() {
+        val array = JSONArray(Arrays.asList(5.5, true))
+        assertEquals(2, array.length().toLong())
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt:523]: toLong

File: AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt [hunk 13]
```diff
@@ -20,507 +20,539 @@
...
  * This black box test was written without inspecting the non-free org.json sourcecode.
  */
-@RunWith(AndroidJUnit4.class)
-@Config(application = EmptyApplication.class)
-public class JSONArrayTest {
-
+@RunWith(AndroidJUnit4::class)
+@Config(application = EmptyApplication::class)
+@KotlinCleanup("Use Kotlin's function instead of Arrays.asList")
+class JSONArrayTest {
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/utils/JSONArrayTest.kt:43]: Cleanup: rename variables named `object`

**Review by david-allison — CHANGES_REQUESTED** [review]

**Review by mikehardy — APPROVED** [review]
Thanks! Verified feedback addressed

**Review by david-allison — APPROVED** [review]

**Pull request merged**